packages/plugin-e2e/src/models/ExplorePage.ts

```diff
@@ -0,0 +1,49 @@
+import { Expect, Locator } from '@playwright/test';
+import { NavigateOptions, PluginTestCtx, RequestOptions } from '../types';
+import { DataSourcePicker } from './DataSourcePicker';
+import { GrafanaPage } from './GrafanaPage';
+import { TimeRange } from './TimeRange';
+
+export class ExplorePage extends GrafanaPage {
+  datasource: DataSourcePicker;
+  timeRange: any;
+  constructor(ctx: PluginTestCtx, expect: Expect<any>) {
+    super(ctx, expect);
+    this.datasource = new DataSourcePicker(ctx, expect);
+    this.timeRange = new TimeRange(ctx, this.expect);
+  }
+
+  async goto(options?: NavigateOptions) {
+    await super.navigate(this.ctx.selectors.pages.Explore.url, options);
+  }
+
+  async getQueryEditorRow(refId: string): Promise<Locator> {
+    //TODO: add new selector and use it in grafana/ui
+    const locator = this.ctx.page.locator('[aria-label="Query editor row"]').filter({
+      has: this.ctx.page.locator(`[aria-label="Query editor row title ${refId}"]`),
+    });
+    await this.expect(locator).toBeVisible();
+    return locator;
+  }
+
+  async runQuery(options?: RequestOptions) {
+    const components = this.ctx.selectors.components;
+    const runButtonTextContent = await this.getByTestIdOrAriaLabel(components.RefreshPicker.runButtonV2).textContent();
+    // if there is a query running, wait for it to finish before starting a new one
+    if (runButtonTextContent === 'Cancel') {
+      await this.ctx.page.waitForResponse((resp) => resp.url().includes('/api/ds/query'), options);
+    }
+
+    const responsePromise = this.ctx.page.waitForResponse((resp) => resp.url().includes('/api/ds/query'), options);
+    try {
+      await this.getByTestIdOrAriaLabel(components.RefreshPicker.runButtonV2).click({
+        timeout: 1000,
+      });
+    } catch (_) {
+      // handle the case when the run button is hidden behind the "Show more items" button
+      await this.getByTestIdOrAriaLabel(components.PageToolbar.item('Show more items')).click();
```

Comment:
not sure how you are handling these magic strings so far, but would it be better to put all these in a file where we can re-use them and have more information related to it.

e.g. `PageToolbar.ShowMoreItemsButton`

This is because the string alone doesn't give context of where is this string supposed to be in the UI and less about being able to re-use or easily update.

Reply:
Yeah this is a huge anti-pattern that I've come across a few times in the grafana/e2e-selectors package. The whole point of this package is to provide a safe way to access these selectors, but with these functions with magic strings we're kind of losing the whole purpose of it. 

I think rewriting all of these selectors in grafana/e2e-selectors will be a too big undertaking for now, but I'll make note in the tasklist to at least change the ones that we use in grafana-e2e (I know they're being used at least two more times).

Reply:
I am not suggesting we move them to e2e/selectors or a whole different package. something as simple as defining the variable at the beginning of the file to avoid this to turn into a  [magic string](https://en.wikipedia.org/wiki/Magic_string)

Reply:
Done